Core/clim-core/presentations/presentation-type-functions.lisp

```diff
@@ -0,0 +1,399 @@
+;;;
+;;; Presentation type functions.
+;;;
+
+(in-package #:clim-internals)
+
+(define-presentation-generic-function %describe-presentation-type
+    describe-presentation-type
+  (type-key parameters options type stream plural-count ))
+
+;;; Support for the default method on describe-presentation-type: if a CLOS
+;;; class has been defined as a presentation type, get description out of the
+;;; presentation type.
+
+(defmethod description ((class standard-class))
+  (let* ((name (class-name class))
+         (ptype-entry (gethash name *presentation-type-table*)))
+    (if ptype-entry
+        (description ptype-entry)
+        (make-default-description name))))
+
+(defun default-describe-presentation-type (description stream plural-count)
+  (if (symbolp description)
+      (setq description (make-default-description (symbol-name description))))
+  (cond ((eql 1 plural-count)
+         (format stream "~:[a~;an~] ~A"
+                   (find (char description 0) "aeiouAEIOU")
+                   description))
+        ((numberp plural-count)
+         (format stream "~D ~A~P" plural-count description plural-count))
+        (plural-count
+         (format stream "~As" description))
+        (t (write-string description stream))))
+
+(define-default-presentation-method describe-presentation-type
+    (type stream plural-count)
+  (with-presentation-type-decoded (name parameters options)
+      type
+    (declare (ignore name parameters))
+    (let ((description (or (getf options :description)
+                           (description (class-of type-key)))))
+      (default-describe-presentation-type description
+                                          stream
+                                          plural-count))))
+
+(defun describe-presentation-type
+    (type &optional (stream *standard-output*) (plural-count 1))
+  (flet ((describe-it (stream)
+           (funcall-presentation-generic-function
+            describe-presentation-type type stream plural-count)))
+    (if stream
+        (describe-it stream)
+        (with-output-to-string (s)
+          (describe-it s)))))
+
+(define-presentation-generic-function
+    %presentation-type-specifier-p
+  presentation-type-specifier-p
+  (type-class type))
+
+(define-default-presentation-method presentation-type-specifier-p (type)
+  t)
+
+(defun presentation-type-specifier-p (object)
+  "Return true if `object' is a valid presentation type specifier,
+otherwise return false."
+  ;; Apparently, this funtion has to handle arbitrary objects.
+  (let ((name (presentation-type-name object)))
+    (when (and (typep name '(or symbol class))
+               (get-ptype-metaclass name))
+      (funcall-presentation-generic-function presentation-type-specifier-p object))))
+
+(define-presentation-generic-function %presentation-typep presentation-typep
+  (type-key parameters object type))
+
+(define-default-presentation-method presentation-typep (object type)
+  (declare (ignore object type))
+  nil)
```

Comment:
This seems to conflict with `(define-presentation-method presentation-typep (object (type t)) …)` in `presentation-defs.lisp`.

Reply:
ditto, default method seems to do the same, but without decoding types.